<launch>
  <arg name="launch_main" default="true" />
  <arg name="json" />
  <arg name="collect_data" default="false" />

  <!-- bin contents data -->
  <node pkg="jsk_2015_05_baxter_apc" type="bin_contents.py" name="bin_contents" respawn="true">
    <param name="json" value="$(arg json)" />
  </node>

  <!-- work order data -->
  <node pkg="jsk_2016_01_baxter_apc" type="work_order.py" name="work_order" respawn="true">
    <rosparam subst_value="true">
      json: $(arg json)
      is_apc2016: true
      gripper: gripper2016
    </rosparam>
  </node>


  <node pkg="jsk_2016_01_baxter_apc" type="output_json_pick.py" name="output_json_pick" respawn="true">
    <rosparam subst_value="true">
        json: $(arg json)
    </rosparam>
  </node>

  <!-- shelf-->
  <arg name="UPPER_SHELF" value="$(find jsk_apc2016_common)/config/bin_upper_shelf.yaml"/>
  <arg name="LOWER_SHELF" value="$(find jsk_apc2016_common)/config/bin_lower_shelf.yaml"/>

  <!-- Publish jsk_apc2016_common/BinInfo from json -->
  <node name="publish_bin_info"
        pkg="jsk_apc2016_common" type="publish_bin_info.py"
        output="screen">
    <param name="json" value="$(arg json)"/>
    <rosparam file="$(arg UPPER_SHELF)" command="load" ns="upper_shelf"/>
    <rosparam file="$(arg LOWER_SHELF)" command="load" ns="lower_shelf"/>
  </node>

  <!-- publish bin bbox -->
  <node name="publish_bin_bbox"
        pkg="jsk_apc2016_common" type="publish_bin_bbox.py"
        output="screen">
    <rosparam file="$(arg UPPER_SHELF)" command="load" ns="upper_shelf"/>
    <rosparam file="$(arg LOWER_SHELF)" command="load" ns="lower_shelf"/>
  </node>

  <!-- publish bin tf -->
  <node pkg="jsk_apc2016_common" type="publish_bin_tf.py"
    name="pub_bin_tf">
    <remap from="~input/bin_info_array" to="/publish_bin_info/bin_array"/>
  </node>

  <!-- parameter -->
  <param name="/left_hand/state" value="" />
  <param name="/left_hand/target_bin" value="" />
  <param name="/right_hand/state" value="" />
  <param name="/right_hand/target_bin" value="" />

  <node pkg="jsk_rqt_plugins" type="rqt_yn_btn" name="rqt_yn_btn" />

  <!-- object verification -->
  <group ns="left_hand">
    <node name="object_verification"
          pkg="jsk_2016_01_baxter_apc" type="apply_bin_contents_hint.py"
          respawn="true">
      <remap from="~input" to="vgg16_object_recognition/output" />
      <param name="json" value="$(arg json)" />
    </node>
  </group>
  <group ns="right_hand">
    <node name="object_verification"
          pkg="jsk_2016_01_baxter_apc" type="apply_bin_contents_hint.py"
          respawn="true">
      <remap from="~input" to="vgg16_object_recognition/output" />
      <param name="json" value="$(arg json)" />
    </node>
  </group>

  <!-- robot main -->
  <node if="$(arg launch_main)" pkg="jsk_2016_01_baxter_apc" type="main.l" name="left_hand"
    args="&quot;(progn (apc-init :larm-controller) (apc-mainloop :larm))&quot;"
    respawn="true" output="screen">
    <env name="DISPLAY" value="" />
    <rosparam subst_value="true">
      collect_data: $(arg collect_data)
    </rosparam>
  </node>
  <node if="$(arg launch_main)" pkg="jsk_2016_01_baxter_apc" type="main.l" name="right_hand"
    args="&quot;(progn (apc-init :rarm-controller) (apc-mainloop :rarm))&quot;"
    respawn="true" output="screen">
    <env name="DISPLAY" value="" />
    <rosparam subst_value="true">
      collect_data: $(arg collect_data)
    </rosparam>
  </node>

  <!-- visualization -->
  <node name="visualize_json"
        pkg="jsk_apc2016_common" type="visualize_pick_json.py"
        args="$(arg json)" />
  <!--
  <node name="visualize_on_rviz"
        pkg="jsk_2015_05_baxter_apc" type="visualize_on_rviz.py">
    <param name="json" value="$(arg json)" />
  </node>
  -->

  <!-- collecting data for sib -->
  <param name="/apc2016_collect_data" value="$(arg collect_data)" />
  <group if="$(arg collect_data)">
    <include file="$(find jsk_2016_01_baxter_apc)/launch/include/collect_sib_data.launch">
      <arg name="hand" value="right"/>
      <arg name="json" value="$(arg json)"/>
    </include>
    <include file="$(find jsk_2016_01_baxter_apc)/launch/include/collect_sib_data.launch">
      <arg name="hand" value="left"/>
      <arg name="json" value="$(arg json)"/>
    </include>
    <node name="data_collection_server_left"
          pkg="jsk_data" type="data_collection_server.py">
      <rosparam>
        save_dir: ~/.ros/jsk_2016_01_baxter_apc/robocup2016
        topics:
          - name: /apply_mask_left_hand/output
            msg_class: sensor_msgs/Image
            savetype: ColorImage
            fname: image.png
          - name: /apply_mask_left_hand/output/mask
            msg_class: sensor_msgs/Image
            savetype: ColorImage
            fname: mask.png
      </rosparam>
    </node>
    <node name="data_collection_server_right"
          pkg="jsk_data" type="data_collection_server.py">
      <rosparam>
        save_dir: ~/.ros/jsk_2016_01_baxter_apc/robocup2016
        topics:
          - name: /apply_mask_right_hand/output
            msg_class: sensor_msgs/Image
            savetype: ColorImage
            fname: image.png
          - name: /apply_mask_right_hand/output/mask
            msg_class: sensor_msgs/Image
            savetype: ColorImage
            fname: mask.png
      </rosparam>
    </node>
  </group>

  <node name="rosconsole_overlay_text"
        pkg="jsk_rviz_plugins" type="rosconsole_overlay_text.py">
    <rosparam>
      nodes:
        - /left_hand
        - /right_hand
    </rosparam>
  </node>

  <node name="team_k_desktop_xdisplay"
        pkg="jsk_baxter_startup" type="xdisplay_image_topic.py"
        args="/visualize_json/output">
  </node>

</launch>
